src/licensedcode/tokenize.py

```diff
@@ -103,17 +105,133 @@ def key_phrase_tokenizer(text, stopwords=STOPWORDS):
+    Returns a list of Spans where in `text`, the `required_phrase` exists,
+    and if it doesn't exist anywhere, returns an empty list.
+    """
+    spans_with_required_phrase = []
+
+    text_tokens = list(index_tokenizer(text=text, preserve_case=preserve_case))
+    required_phrase_tokens = list(index_tokenizer(text=required_phrase, preserve_case=preserve_case))
+    required_phrase_first_token = required_phrase_tokens[0]
+
+    # Initial check to see if all tokens in the required phrase are present
+    if all([
+        required_phrase_token in text_tokens
+        for required_phrase_token in required_phrase_tokens
+    ]):
+        start_positions = [
+            i
+            for i, x in enumerate(text_tokens)
+            if x == required_phrase_first_token
+        ]
+
+        for start_pos in start_positions:
+            end_pos = start_pos + len(required_phrase_tokens)
+
+            if end_pos <= len(text_tokens) and text_tokens[start_pos:end_pos] == required_phrase_tokens:
+                spans_with_required_phrase.append(
+                    Span(start_pos, end_pos-1)
+                )
+
+    return spans_with_required_phrase
+
+
+def get_ignorable_spans(rule):
+    """
+    Return a list of Spans which are the positions in the rule text there
+    are ignorable URLs or referenced filenames.
+    """
+    ignorable_spans = []
+    ignorables = rule.referenced_filenames + rule.ignorable_urls
+    for ignorable in ignorables:
+        ignorable_spans.extend(
+            return_spans_for_required_phrase_in_text(
+                text=rule.text,
+                required_phrase=ignorable,
+                preserve_case=True,
+            )
+        )
+
+    return ignorable_spans
+
+
+def get_non_overlapping_spans(old_required_phrase_spans, new_required_phrase_spans):
+    """
+    Given two list of spans `old_required_phrase_spans` and `new_required_phrase_spans`,
+    return all the spans in `new_required_phrase_spans` that do not overlap with any
+    of the spans in `old_required_phrase_spans`.
+
+    The list of spans `old_required_phrase_spans` contains all the spans of required
+    phrases or ignorables already present in a rule text, and the other list of spans
+    `new_required_phrase_spans` contains the proposed new required phrases.
+    """
+    if not old_required_phrase_spans:
+        return new_required_phrase_spans
+    
+    for new_span in new_required_phrase_spans:
+        if not any(
+            old_span.overlap(new_span) != 0
+            for old_span in old_required_phrase_spans
+        ):
+            yield new_span
+
+
+def combine_tokens(token_tuples):
+    """
+    Returns a string `combined_text` combining token tuples from the list `token_tuples`,
+    which are token tuples created by the tokenizer functions.
+    """
+    combined_text = ''
+    
+    for token_tuple in token_tuples:
+        _value, token = token_tuple
+        combined_text += token
+        
+    return combined_text
```

Comment:
What about this, without intermediate accumulation of strings?
```suggestion
    return ''.join(token for _, token in token_tuples)
```